pingcap/tidb
# executor: error processing refactor for load data

### What problem does this PR solve? 

there are two different modifications(maybe they should not be considered together?)

1. remove old logic in load data error processing,  `commitWork` function will commit or rollback txn. Former implementation will make query processing panic since commit fail and `txn` is invalid.
Ealier  tidb version still has this problem(2.1 3.0), you will get `lost connection` error if doing sth like ddl between `load data`

2. `doCommit` function defer will make txn `changeToInvalid`, so if `commit` failed and `stmtRollBack` is called, `txnState` has nil `transaction` field,  which will cause `ConfirmAssertions`  operation report panic error "invalid memory nil"


some unit-test or tidb-test will be added for this situation

```
enable commit fail point in `commitOneTask` function
do load data query

client expected like:
ERROR 1105 (HY000): Information schema is changed. [try again later]
not `lost connection` and no panic error log
```

### What is changed and how it works?


### Check List 

Tests 


 - Manual test (add detailed scripts or steps below)
```
enable commit fail point in `commitOneTask` function
do load data query

client expected like:
ERROR 1105 (HY000): Information schema is changed. [try again later]
not `lost connection` and no panic error log
```

Code changes



Side effects



Related changes



Release note
